### ISSUE TITLE
vncviewer loses keyboard focus after unlocking the screen

### ISSUE THREAD

**Issue by rtwruck** [TigerVNC/tigervnc#1150]
**Describe the bug**
When running in full screen mode, after unlocking a VT that was locked by `light-locker`, `vncviewer` still passes mouse events to the server but accepts no keyboard input. No keys are sent to the server and the menu does not appear. It's impossible to leave full screen mode and `vncviewer` has to be killed since it's impossible to exit the viewer using the mouse only.

When not running in full screen mode while the screen gets locked, the problem does not appear.

**To Reproduce**
1. Connect `vncviewer` to a server
2. Enable full screen mode
3. Wait for `light-locker` to lock the screen
4. Unlock the screen
5. Press F8

**Expected behavior**
Menu opens - but it doesn't.

**Client (please complete the following information):**
 - OS: Linux (Xubuntu 20.04.1)
 - VNC client: TigerVNC Viewer 64-bit
 - VNC client version: 1.11.0
 - Client downloaded from: GitHub (v1.11.0), same behaviour with Ubuntu package (v1.10.1)

**Server (please complete the following information):**
 - OS: Linux (Xubuntu 20.04.1)
 - VNC server: Xvnc TigerVNC 64-bit
 - VNC server version: 1.10.0
 - Server downloaded from: Ubuntu package
 - Server was started using: vncserver

**Additional context**
Client window manager: `xfwm4`

**Comment by rtwruck** [comment]
Disabling `numlockx` (since it seems to mess with the keyboard to enable num lock for the lock screen) did not help.

**Comment by ppira** [comment]
Does "vncviewer -FullScreenSystemKeys=0" make a difference?

**Comment by rtwruck** [comment]
Actually YES, that works.
Unfortunately though, Alt-Tab was one of the reasons I used full screen mode and that stopped working - as documented.

**Comment by CendioOssman** [comment]
Something broken with regard to keyboard grabbing and focus events in that case. Do you know if you are running a standard `Xorg` on the client, or something like `Wayland`?

Does it break with other screen savers or desktop environments?

**Comment by rtwruck** [comment]
It's plain Xorg:
`/usr/lib/xorg/Xorg -core :0 -seat seat0 -auth /var/run/lightdm/root/:0 -nolisten tcp vt7 -novtswitch`

Same problem when using xscreensaver.

Will try with GNOME / gdm3 / gnome-screensaver next.

**Comment by CendioOssman** [comment]
I can reproduce the issue here, but it smells like a X server bug right now. Somehow the screen saver is able to grab focus away from us, which is odd given that we have a grab. What really screws things up is that we don't get that focus back properly once the screen saver is done.

I'll need to spend more time on this to understand what is happening.

**Comment by CendioOssman** [comment]
After more digging this is some issue with the screensavers (and possibly the window manager). They explicitly remove focus from TigerVNC in order to lock the screen. But then they fail to properly restore things and focus ends up getting wedged.

I reported it to the xscreensaver maintainer, but he refuses to look at it so the verdict there is simply "don't run xscreensaver if you use TigerVNC".

For Xfce I've filed a bug here:

https://gitlab.xfce.org/apps/xfce4-screensaver/-/issues/70

Hopefully they can provide some fix.

Unfortunately there isn't anything more we can do from our end, so I'll go ahead and close this issue.
